tests/link/test_numba.py

```diff
@@ -878,6 +878,7 @@ def test_Reshape_scalar():
     )
 
 
+@pytest.mark.xfail(reason="Fails with Nones")
```

Comment:
Couldn't figure out how to get Numba to work with Nones

Reply:
It's a compilation issue not just a `compare_numba_and_py`.

```python
import aesara
import aesara.tensor as at
import numpy as np
x = at.matrix("x")
y = at.specify_shape(x, (1, None))
f = aesara.function([x], y, mode="NUMBA")
f(np.zeros((1, 2)))
```

<details>
 <summary>Traceback</summary>

```
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 202, in streamline_default_f
    thunk()
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/basic.py", line 664, in thunk
    outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 482, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 423, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<class 'zip'>) found for signature:
 
 >>> zip(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload of function 'zip': File: numba/core/typing/builtins.py: Line 1019.
    With argument(s): '(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))':
   No match.
During: resolving callee type: Function(<class 'zip'>)
During: typing of call at /home/ricardo/Documents/Projects/aesara/aesara/link/numba/dispatch/basic.py (589)
File "aesara/link/numba/dispatch/basic.py", line 589:
    def specifyshape(x, *shape):
        for xs, s in zip(x.shape, shape):
        ^
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
File "../../../../../tmp/tmphikq3lce", line 3:
def numba_funcified_fgraph(x):
    auto_59 = specifyshape(x, auto_55, NoneConst)
    ^
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3441, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-462a1a0bd99b>", line 7, in <module>
    f(np.zeros((1, 2)))
  File "/home/ricardo/Documents/Projects/aesara/aesara/compile/function/types.py", line 964, in __call__
    self.fn()
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 206, in streamline_default_f
    raise_with_op(fgraph, node, thunk)
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 538, in raise_with_op
    raise exc_value.with_traceback(exc_trace)
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/utils.py", line 202, in streamline_default_f
    thunk()
  File "/home/ricardo/Documents/Projects/aesara/aesara/link/basic.py", line 664, in thunk
    outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 482, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/ricardo/Documents/Projects/aesara/venv/lib/python3.8/site-packages/numba/core/dispatcher.py", line 423, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<class 'zip'>) found for signature:
 
 >>> zip(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload of function 'zip': File: numba/core/typing/builtins.py: Line 1019.
    With argument(s): '(UniTuple(int64 x 2), StarArgTuple(readonly array(int8, 0d, C), none))':
   No match.
During: resolving callee type: Function(<class 'zip'>)
During: typing of call at /home/ricardo/Documents/Projects/aesara/aesara/link/numba/dispatch/basic.py (589)
File "aesara/link/numba/dispatch/basic.py", line 589:
    def specifyshape(x, *shape):
        for xs, s in zip(x.shape, shape):
        ^
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
During: resolving callee type: type(CPUDispatcher(<function numba_funcify_SpecifyShape.<locals>.specifyshape at 0x7f11594b48b0>))
During: typing of call at /tmp/tmphikq3lce (3)
File "../../../../../tmp/tmphikq3lce", line 3:
def numba_funcified_fgraph(x):
    auto_59 = specifyshape(x, auto_55, NoneConst)
    ^
Apply node that caused the error: DeepCopyOp(SpecifyShape.0)
Toposort index: 1
Inputs types: [TensorType(float64, (1, None))]
Inputs shapes: [(1, 2)]
Inputs strides: [(16, 8)]
Inputs values: [array([[0., 0.]])]
Outputs clients: [['output']]
HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.
```
</details>

Reply:
I tried a bunch of things, include creating and index of non-None inputs outside and inside the function, but nothing worked for me. Just different errors.

Reply:
Is this still an issue?

Reply:
Yes. Kaustubh wrote this implementation, but it seems pretty involved:

```python
@numba_funcify.register(SpecifyShape)
def numba_funcify_SpecifyShape(op, node, **kwargs):
    
    num_shape_args = len(node.inputs) - 1
    input_names = ["shape_" + str(i) for i in range(num_shape_args)]
    shape_indices = [f"x.shape[{i}]" for i in range(num_shape_args)]

    func_body = ""
    for _input_name, _shape_index in zip(input_names, shape_indices):
        func_body = func_body + f"""
    if {_input_name} is not None:
        assert {_shape_index} == {_input_name}
"""

    func_template = f"""
def specifyshape(x, {create_arg_string(input_names)}):
    {func_body}
    return x
"""
    specifyshape = compile_function_src(
        func_template, "specifyshape", globals()
    )
    return numba_njit(specifyshape)
```

Reply:
I suggested instead doing o numba-specific rewrite where we replace `None` by `-1` just before compilation, which would get rid of the issues with mixed arg types that Numba seems to reject. I don't have a strong preference for either option. What do you think?